### PR TITLE
Added support for left_pad to HookedTransformer

### LINKED ISSUE #240: [Proposal] Add support for padding/masking to the attention computation (Minor feature request)
### Proposal 

Models like BERT allow padding to the attention pattern, where the attention pattern is always zero for <PAD> tokens, ie the model cannot look at those tokens. This can also be done by explicitly passing in a Boolean attention mask, for models which do not have a separate PAD and END_OF_TEXT/START_OF_TEXT (EOS/BOS) tokens like GPT-2. 

### Motivation

This is mildly useful to have for even GPT-style models, as it means that if you eg input a batch of prompts of different lengths, they can be padded on the left, so that `logits[:, -1, :]` gets you the final token of each prompt, rather than logits[:, -1, :] being meaningless for shorter prompts.

I particularly want this to be part of exploratory analysis demo, so that when people input individual prompts it's harder to shoot yourself in the foot by having them be of different lengths. 

### Alternatives

The main alternative is just to do padding manually, or massage your prompts to all have the same length. 

### Additional context

This would also need to change absolute positional embeddings to use the attention mask, so that the first real token in each prompt gets the 0th positional embedding (it'll break things otherwise). I would implement this by having an attention mask object that gets optionally created by `to_tokens` if you give it a `left_pad=True` option, or possibly have a `to_tokens_left_pad` option, and which gets passed to `pos_embed` and to `attn`. Relative positional embedding schemes *should* be unaffected. 

This isn't that big a deal, just a mildly nice feature to have.

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/Farama-Foundation/MiniGrid/issues) in the repo (**required**)

### PR DESCRIPTION
**Pull request by JayBaileyCS**
# Description

Replaces PR #291 because it waited too long and is now too big and messy.

* Added keyword argument left_pad (default False to avoid breaking anything) to HookedTransformer. This keyword argument sets the tokenizer's "padding_side" argument to left if True.
* Added attention mask for when padding_side is left to ensure we don't attend to padded tokens.
* Added test to support this change. (This doesn't need to be tested much because it leverages HuggingFace's implementation - we just have to verify it actually runs in forward)
* Added this change to the Exploratory Analysis Demo.
* Fixed a test in test_utils that was failing by enforcing dtype=np.int64

Fixes #240 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

**Review by jbloomAus — CHANGES_REQUESTED** [review]
Hey Jay,

Looks good, can you please write a test for / visualize an attention map with padding tokens to show that it's working. I can see the padding tokens are on the write side but there's not test for the mask being correct. 

Cheers,

**Comment by JayBaileyCS** [comment]
Having looked further into this, it seems like it would be a more major undertaking than is worth for a "nice-to-have". The additive attention mask (where Joseph advised me to add it) is part of the Attention block. These blocks are made on construction of the transformer, so you can't pass in the input tokens to generate the mask. It would require rethinking the whole thing, which could be done if this was crucial, but in this case it isn't.
